src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -74,87 +157,98 @@ class FunctionBuilder[F >: Null](parameterTypeInfo: Array[MaybeGenericTypeInfo[_
   cn.superName = "java/lang/Object"
   cn.interfaces.asInstanceOf[java.util.List[String]].add("java/io/Serializable")
 
-  def descriptor: String = s"(${ parameterTypeInfo.map(_.base.name).mkString })${ returnTypeInfo.base.name }"
+  var methods: mutable.ArrayBuffer[MethodBuilder] = new mutable.ArrayBuffer[MethodBuilder](16)
 
-  val mn = new MethodNode(ACC_PUBLIC, "apply", descriptor, null, null)
+  val apply = new MethodBuilder(this, "apply", parameterTypeInfo.map(_.base), returnTypeInfo.base)
   val init = new MethodNode(ACC_PUBLIC, "<init>", "()V", null, null)
   // FIXME why is cast necessary?
-  cn.methods.asInstanceOf[util.List[MethodNode]].add(mn)
   cn.methods.asInstanceOf[util.List[MethodNode]].add(init)
 
   init.instructions.add(new IntInsnNode(ALOAD, 0))
   init.instructions.add(new MethodInsnNode(INVOKESPECIAL, Type.getInternalName(classOf[java.lang.Object]), "<init>", "()V", false))
   init.instructions.add(new InsnNode(RETURN))
 
-  val start = new LabelNode
-  val end = new LabelNode
+  if (parameterTypeInfo.exists(_.isGeneric) || returnTypeInfo.isGeneric) {
+    val generic = new MethodBuilder(this, "apply", parameterTypeInfo.map(_.generic), returnTypeInfo.generic)
+    methods.append(generic)
+    generic.emit(
+      new Code[Unit] {
+        def emit(il: Growable[AbstractInsnNode]) {
+          returnTypeInfo.castToGeneric(
+            apply.invoke(parameterTypeInfo.zipWithIndex.map { case (ti, i) =>
+              ti.castFromGeneric(generic.getArg(i + 1)(ti.generic))
```

Comment:
why i + 1?

Reply:
the args are 1 indexed because getArg(0) is the class that the method is defined on. this is how FunctionBuilder as things set up, and I didn't see much reason to change it since an "INVOKE" bytecode op takes (object, arg1, arg2, arg3, ...) off the stack, and the first local variable stores the object.

Reply:
Ah, got it. Thanks!